packages/protocol/chainmanager/manager.go

```diff
@@ -77,23 +85,74 @@ func (m *Manager) ProcessCommitment(commitment *commitment.Commitment) (isSolid
+	return
 }
 
-func (m *Manager) registerCommitment(commitment *commitment.Commitment) (isSolid bool, chain *Chain, wasForked bool, chainCommitment *Commitment, commitmentPublished bool) {
-	m.commitmentEntityMutex.Lock(commitment.ID())
-	defer m.commitmentEntityMutex.Unlock(commitment.ID())
+func (m *Manager) ProcessCommitmentFromSource(commitment *commitment.Commitment, source identity.ID) (isSolid bool, wasForked bool, chain *Chain) {
+	_, isSolid, wasForked, chainCommitment := m.processCommitment(commitment)
 
-	chainCommitment, created := m.Commitment(commitment.ID(), true)
+	if wasForked {
+		m.handleFork(chainCommitment, chainCommitment.Chain(), source)
+	}
 
-	if !chainCommitment.PublishCommitment(commitment) {
-		return chainCommitment.IsSolid(), chainCommitment.Chain(), false, chainCommitment, false
+	return isSolid, wasForked, chainCommitment.Chain()
+}
+
+func (m *Manager) ProcessCandidateCommitment(commitment *commitment.Commitment) (isSolid bool, wasForked bool, chain *Chain) {
+	_, isSolid, wasForked, chainCommitment := m.processCommitment(commitment)
+	return isSolid, wasForked, chainCommitment.Chain()
+}
+
+func (m *Manager) ProcessCommitment(commitment *commitment.Commitment) (isSolid bool, wasForked bool, chain *Chain) {
+	isNew, isSolid, wasForked, chainCommitment := m.processCommitment(commitment)
+
+	if !isNew || wasForked {
+		if err := m.switchMainChainToCommitment(chainCommitment); err != nil {
+			panic(err)
+		}
 	}
 
-	if !created {
-		m.Events.MissingCommitmentReceived.Trigger(chainCommitment.ID())
+	return isSolid, false, chainCommitment.Chain()
+}
+
+func (m *Manager) handleFork(commitment *Commitment, chain *Chain, source identity.ID) {
+	commitmentsStorage := m.commitmentsByForkingPoint.Get(commitment.ID().Index(), true)
+
+	forkingPoint, err := m.forkingPointAgainstMainChain(commitment)
+	if err != nil {
+		return
+	}
+
+	if forkingPoint == nil {
+		return
+	}
+
+	// Do not trigger another event for the same forking point.
+	if commitmentsStorage.Has(forkingPoint.ID()) {
+		return
 	}
 
+	// The forking point should be at least 3 epochs in the past w.r.t this commitment index.
+	//TODO: replace 3 with values depending on the configuration
+	latestChainCommitment := chain.LatestCommitment()
+	if latestChainCommitment.ID().Index()-forkingPoint.ID().Index() < 3 {
+		return
+	}
+
+	// commitment.Index is the index you request attestations for.
+	event := &ForkDetectedEvent{
+		Source:                       source,
+		Commitment:                   latestChainCommitment.Commitment(),
+		ForkingPointAgainstMainChain: forkingPoint.Commitment(),
+	}
+	commitmentsStorage.Set(forkingPoint.ID(), event)
+	m.Events.ForkDetected.Trigger(event)
+}
+
+func (m *Manager) registerCommitment(commitment *commitment.Commitment) (isNew bool, isSolid bool, wasForked bool, chainCommitment *Commitment) {
+	m.commitmentEntityMutex.Lock(commitment.ID())
+	defer m.commitmentEntityMutex.Unlock(commitment.ID())
```

Comment:
Can this lead to a deadlock if another goroutine is currently in `processCommitment` to propagate solidity to the children. There we hold the lock of the parent in `processCommitment` and update the child in `propagateSolidity` where we lock the child. Here we do it the opposite way.